optimizely/event/dispatcher.go

```diff
@@ -54,10 +55,72 @@ func (*HTTPEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
 		if resp.StatusCode == 204 {
 			success = true
 		} else {
-			fmt.Printf("http.Post invalid response %d", resp.StatusCode)
+			dispatcherLogger.Error(fmt.Sprintf("http.Post invalid response %d", resp.StatusCode), nil)
 			success = false
 		}
 	}
 	callback(success)
+}
+
+// QueueEventDispatcher is a queued version of the event dispatcher that queues, returns success, and dispatches events in the background
+type QueueEventDispatcher struct {
+	eventQueue     Queue
+	eventFlushLock sync.Mutex
+	dispatcher     *HTTPEventDispatcher
+}
+
+// DispatchEvent queues event with callback and calls flush in a go routine.
+func (ed *QueueEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
+
+	ed.eventQueue.Add(event)
+	callback(true)
+	go func() {
+		ed.flushEvents()
+	}()
+}
+
+// flush the events
+func (ed *QueueEventDispatcher) flushEvents() {
+
+	ed.eventFlushLock.Lock()
+
+	defer func(){
+		ed.eventFlushLock.Unlock()
+	}()
+
+	for ed.eventQueue.Size() > 0 {
+		items := ed.eventQueue.Get(1)
+		if len(items) == 0 {
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		event, ok := items[0].(LogEvent)
+		if !ok {
+			// remove it
+			dispatcherLogger.Error("invalid type passed to event dispatcher", nil)
+			ed.eventQueue.Remove(1)
+			continue
+		}
+
+		ed.dispatcher.DispatchEvent(event, func(success bool) {
+			if success {
+				ed.eventQueue.Remove(1)
+			} else {
+				// we failed.  Sleep 5 seconds and try again.  Or, should we exit and try the next time a log event is added?
+				time.Sleep(5 * time.Second)
```

Comment:
should we log this ? i think a warning  would be useful here

Reply:
Use a constant for this magic number